src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/RdsHostUtils.java

```diff
@@ -0,0 +1,304 @@
+      return RDS_INSTANCE;
+    } else {
+      return OTHER;
+    }
+  }
+
+  public RdsHost getRdsHost(PropertySet propertySet, HostInfo hostInfo) throws SQLException {
+    final String hostName = hostInfo.getHost();
+    final int hostPort = hostInfo.getPort();
+    final RdsUrlType rdsUrlType = getUrlType(hostName);
+    final String clusterId = getClusterId(propertySet, rdsUrlType, hostName, hostPort);
+
+    String clusterInstanceTemplateHost = hostName;
+    if (rdsUrlType.isRds()) {
+      clusterInstanceTemplateHost = getRdsInstanceHostPattern(hostName);
+      if (clusterInstanceTemplateHost == null) {
+        this.logger.logError(Messages.getString("ConnectionStringTopologyProvider.5"));
+        throw new SQLException(Messages.getString("ConnectionStringTopologyProvider.5"));
+      }
+    }
+
+    final HostInfo clusterInstanceTemplate = getClusterInstanceTemplate(propertySet, hostInfo, clusterInstanceTemplateHost, hostPort);
+    return new RdsHost(rdsUrlType, clusterId, clusterInstanceTemplate, hostName, hostInfo.getPort());
+  }
+
+  public RdsHost getRdsHostFromHostPattern(PropertySet propertySet, HostInfo hostInfo, String pattern) throws SQLException {
+    final ConnectionUrlParser.Pair<String, Integer> pair = ConnectionUrlParser.parseHostPortPair(pattern);
+    if (pair == null) {
+      // "Invalid value for the 'clusterInstanceHostPattern' configuration setting - the value could not be parsed"
+      throw new SQLException(Messages.getString("ConnectionStringTopologyProvider.1"));
+    }
+
+    final String hostName = pair.left;
+    final RdsUrlType rdsUrlType = getUrlType(hostName);
+    validateHostPatternSetting(hostName, rdsUrlType);
+
+    final int hostPort =
+        pair.right != RdsHost.NO_PORT ? pair.right : hostInfo.getPort();
+
+    final String clusterId = getClusterId(propertySet, rdsUrlType, hostName, hostPort);
+    final HostInfo clusterInstanceTemplate = getClusterInstanceTemplate(propertySet, hostInfo, hostName, hostPort);
+    return new RdsHost(rdsUrlType, clusterId, clusterInstanceTemplate, hostName, hostPort);
+  }
+
+  private void validateHostPatternSetting(String hostName, RdsUrlType rdsUrlType) throws SQLException {
+    if (!isDnsPatternValid(hostName)) {
+      // "Invalid value for the 'clusterInstanceHostPattern' configuration setting - the host pattern must contain a '?'
+      // character as a placeholder for the DB instance identifiers of the instances in the cluster"
+      this.logger.logError(Messages.getString("ConnectionStringTopologyProvider.2"));
+      throw new SQLException(Messages.getString("ConnectionStringTopologyProvider.2"));
+    }
+
+    if (RDS_PROXY.equals(rdsUrlType)) {
+      // "An RDS Proxy url can't be used as the 'clusterInstanceHostPattern' configuration setting."
+      this.logger.logError(Messages.getString("ConnectionStringTopologyProvider.3"));
+      throw new SQLException(Messages.getString("ConnectionStringTopologyProvider.3"));
```

Comment:
I propose we store `Messages.getString` to a variable, and create a method for this log and throw logic, since it is used a few times

Reply:
Good idea, sounds good!